### PR TITLE
Introduce build time flag property to disable the cache extension

### LINKED ISSUE #11138: Introduce build time flag property to disable the cache extension
**Description**
This is originally an idea from @seseso in #8140 (see [here](https://github.com/quarkusio/quarkus/issues/8140#issuecomment-667349486)). Many Quarkus extensions can be disabled using a flag property while this is not currently possible for the cache extension. As a consequence, one has to remove/comment all caching annotations to disable the extension.

### PR DESCRIPTION
**Pull request by gwenneg**
Fixes #11138.

### CODE REVIEW

File: extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java [hunk 1]
```diff
@@ -113,4 +114,13 @@ void recordCachesBuild(CombinedIndexBuildItem combinedIndex, BeanContainerBuildI
         }
         return cacheNames;
     }
+
+    private static class IsEnabled implements BooleanSupplier {
```
Comment (geoand) [inline extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java:118]: Minor nitpick: Can we call give this a more descriptive name, something like `CacheEnabled` or something?
That would make it more consistent with the the existing uses of `onlyIf`

Comment (gwenneg) [reply extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheProcessor.java]: That's a good point, I just applied your suggestion. Thanks for reviewing this!

**Comment by jaikiran** [comment]
Hello @gwenneg, if this config set to "disable this extension", then would adding this extension to the application's dependency (`pom.xml` for example) be of any value? In other words, can't the extension be disabled by just removing it from the application's dependency?

**Comment by geoand** [comment]
@jaikiran this kind of configuration usually proves helpful in testing scenarios

**Comment by jaikiran** [comment]
> @jaikiran this kind of configuration usually proves helpful in testing scenarios

Hello @geoand, on the other hand, wouldn't this mean that every single build step method of every build step processor that corresponds to this extension will now have to add:

```
    @BuildStep(onlyIf = CacheEnabled.class)
```
Perhaps, if we do need something like this, maybe this should be some generic config which can then be implemented in the build step processor invocation layer where with a single check it can decide if the processor and its build steps are relevant or not depending on that enabled/disabled config?

**Comment by geoand** [comment]
Interesting idea @jaikiran. 

Are you thinking something `@BuildStep(onlyIfProperty ="some.value")`?

**Comment by gwenneg** [comment]
> Hello @gwenneg, if this config set to "disable this extension", then would adding this extension to the application's dependency (`pom.xml` for example) be of any value? In other words, can't the extension be disabled by just removing it from the application's dependency?

This PR started from a Quarkus user who needed to disable the extension temporarily for testing purposes. This cannot be done by removing the extension from the application `pom.xml` since it would also mean that all caching annotations must be removed from the source code.

**Review by geoand — APPROVED** [review]

**Comment by gwenneg** [comment]
> > @jaikiran this kind of configuration usually proves helpful in testing scenarios
> 
> Hello @geoand, on the other hand, wouldn't this mean that every single build step method of every build step processor that corresponds to this extension will now have to add:
> 
> ```
>     @BuildStep(onlyIf = CacheEnabled.class)
> ```
> 
> Perhaps, if we do need something like this, maybe this should be some generic config which can then be implemented in the build step processor invocation layer where with a single check it can decide if the processor and its build steps are relevant or not depending on that enabled/disabled config?

For this specific case, I didn't disable all build steps. The one providing the FeatureBuildItem instance is still active since the extension is still part of the application (even if it is disabled). Also, when the programmatic API of this extension will be there, some other build steps will certainly remain active while the annotations caching API might be disabled.

**Comment by gwenneg** [comment]
@geoand Now that I wrote the previous comment, I am wondering if the property included in this PR should contain a part indicating that it will disable the annotations... Something like `quarkus.cache.annotations.enabled` or `quarkus.cache.annotation-api.enabled`.

**Comment by geoand** [comment]
@gwenneg I don't think that is necessary. Maybe just update the javadoc of the property to mention that the caching annotations will have no effect

**Comment by gwenneg** [comment]
@geoand I did a small `CacheConfig` Javadoc update to include your suggestion.

**Comment by geoand** [comment]
Looks good, thanks!

**Review by gsmet — APPROVED** [review]
I'm a bit torn on that one as I wonder if it would not make more sense to just disable the interceptors and keep the rest of the logic running.

But maybe it could be an issue if it was using an external cache service which we might support at some point.

Merging anyway.

**Comment by gwenneg** [comment]
@gsmet I also hesitated between disabling the entire extension logic at build time or disabling the interceptors logic only at run time. But since this PR started from a user who wanted to compare performances with and without the cache extension, disabling it at build time seemed more relevant to me. Anyway, this could easily be changed in the future.

Thanks for merging!

**Pull request merged**